cpp/src/io/csv/reader_impl.cu

```diff
@@ -346,194 +351,215 @@ table_with_metadata reader::impl::read(size_t range_offset,
+          break;
         }
       }
-    }
-    if (filtered_count != count) {
-      it_end = std::remove_if(
-        it_begin, it_end, [](uint64_t pos) { return (pos == static_cast<uint64_t>(-1)); });
-    }
-  }
-
-  // Exclude the rows that are to be skipped from the start
-  if (skip_rows > 0 && skip_rows < std::distance(it_begin, it_end)) { it_begin += skip_rows; }
-
-  // Exclude the rows outside of requested range
-  if (range_size != 0) {
-    auto it = it_end - 1;
-    while (it >= it_begin && *it > static_cast<uint64_t>(range_size)) { --it; }
-    if ((it + 2) < it_end) { it_end = it + 2; }
-  }
-
-  // Exclude the rows without data
-  if (opts.skipblanklines || opts.comment != '\0') {
-    const auto newline  = opts.skipblanklines ? opts.terminator : opts.comment;
-    const auto comment  = opts.comment != '\0' ? opts.comment : newline;
-    const auto carriage = (opts.skipblanklines && opts.terminator == '\n') ? '\r' : comment;
-
-    it_end = std::remove_if(it_begin, it_end, [=, &h_data](uint64_t pos) {
-      return ((pos != h_size) &&
-              (h_data[pos] == newline || h_data[pos] == comment || h_data[pos] == carriage));
-    });
-  }
-
-  // Exclude the rows before the header row (inclusive)
-  if (std::distance(it_begin, it_end) > 1) {
-    if (args_.header == -1) {
-      header.assign(h_data + *(it_begin), h_data + *(it_begin + 1));
+      if (num_rows > 0 && num_row_offsets > static_cast<size_t>(num_rows)) {
+        // Got the desired number of rows
+        break;
+      }
     } else {
-      header.assign(h_data + *(it_begin + args_.header), h_data + *(it_begin + args_.header + 1));
-      it_begin += args_.header + 1;
+      // Discard data (all rows below skip_rows), keeping one character for history
+      size_t discard_bytes = std::max(data_.size(), sizeof(char)) - sizeof(char);
+      if (discard_bytes != 0) {
+        data_.erase(data_.begin(), data_.begin() + discard_bytes);
+        buffer_pos += discard_bytes;
+      }
     }
-  }
-
-  // Exclude the rows that exceed past the requested number
-  if (num_rows >= 0 && num_rows < std::distance(it_begin, it_end)) {
-    it_end = it_begin + num_rows + 1;
-  }
-
-  // Exclude the rows that are to be skipped from the end
-  if (skip_end_rows > 0 && skip_end_rows < std::distance(it_begin, it_end)) {
-    it_end -= skip_end_rows;
-  }
-
-  const uint64_t offset_start = *it_begin;
-  const uint64_t offset_end   = *(it_end - 1);
+    pos = target_pos;
+  } while (pos < h_size);
+}
 
-  // Copy out the row starts to use for row-column data parsing
-  if (offset_start != offset_end) {
-    if (offset_start != 0) {
-      for (auto it = it_begin; it != it_end; ++it) { *it -= offset_start; }
+size_t reader::impl::parse_csv_header(const char *h_data, size_t h_size)
+{
+  using namespace cudf::io::csv::gpu;
+  size_t header_start = 0;
+  size_t header_end   = 0;
+  size_t data_start   = 0;
+  cudf::size_type row = 0;
+  int ctx             = ROW_CTX_NONE;
```

Comment:
It might be helpful to call this a state or something to disambiguate it from the 20-bit context (row count + id/row_ctx) as used elsewhere.

Reply:
Ok, went with "state". Oh wait, I just realized I renamed the wrong variable in csv_gpu.cpp. Gimme a sec.

Reply:
Done. "state" definitely makes it clear that it's not quite the same thing as in the kernel.